src/modules/ecdsa_sign_to_contract/main_impl.h

```diff
@@ -0,0 +1,46 @@
+/**********************************************************************
+ * Copyright (c) 2019 Marko Bencun, Jonas Nick                        *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef SECP256K1_MODULE_ECDSA_SIGN_TO_CONTRACT_MAIN_H
+#define SECP256K1_MODULE_ECDSA_SIGN_TO_CONTRACT_MAIN_H
+
+#include "include/secp256k1_ecdsa_sign_to_contract.h"
+
+int secp256k1_ecdsa_s2c_sign(const secp256k1_context *ctx, secp256k1_ecdsa_signature *signature, secp256k1_s2c_opening *s2c_opening, const unsigned char *msg32, const unsigned char *seckey, const unsigned char* s2c_data32) {
+    return secp256k1_ecdsa_sign_helper(ctx, signature, s2c_opening, msg32, seckey, s2c_data32, NULL, NULL);
+}
+
+int secp256k1_ecdsa_s2c_verify_commit(const secp256k1_context* ctx, const secp256k1_ecdsa_signature *sig, const unsigned char *data32, const secp256k1_s2c_opening *opening) {
+    secp256k1_pubkey commitment;
+    secp256k1_ge commitment_ge;
+    unsigned char x_bytes1[32];
+    unsigned char x_bytes2[32];
+    secp256k1_scalar sigr, sigs;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(sig != NULL);
+    ARG_CHECK(data32 != NULL);
+    ARG_CHECK(opening != NULL);
+    ARG_CHECK(secp256k1_s2c_commit_is_init(opening));
+
+    if (!secp256k1_ec_commit(ctx, &commitment, &opening->original_pubnonce, data32, 32)) {
+        return 0;
+    }
+
+    /* Check that sigr (x coordinate of R) matches the x coordinate of the commitment. */
```

Comment:
> nit: stress somehow that sigr is a scalar and the x coordinate of the commitment is a field element.

Will do, though I thought that was clear as it is the same way with regular ecdsa verification.

> Note that sig[0:32] can be greater than group order and therefore sigr will be reduced. But in that case ecdsa_verify doesn't pass

That is not true, is it? ecdsa_verify can still pass in this case, as the condition is ` sig_r == x (mod n)` (the field element `x` is also reduced module `n`).

In the current code a few lines down, there actually might be a "mistake": we memory-compare the byte representation of the *scalar* `sigr` and the *field element* `x`, without first reducing `x modulo n`, which might fail for actually valid commitments with very low (negligible) probability. 

Shouldn't this be implemented the same way as the actual `secp256k1_ecdsa_verify()`? Basically convert the field element `x` to a scalar and use `secp256k1_scalar_eq()` instead of `memcmp`, like here:

https://github.com/bitcoin-core/secp256k1/blob/074ab582ddb9085611cba8a0495828dadb645345/src/ecdsa_impl.h#L236-L238

Reply:
Generally a bit confused, I don't see "modulo n" reduction happening in signing nor verification despite it being part of the algorithm as described here:

https://en.wikipedia.org/wiki/Elliptic_Curve_Digital_Signature_Algorithm

Reply:
Added one commit to extend the comment, and one to change to `secp256k1_scalar_eq()` to properly compare the values modulo `n`. The second comment about the reduced security will follow soon.

Reply:
> More importantly, there needs to be a sentence explaining why comparing the x coordinate is sufficient. For example, because there's only two y coordinates per x coordinate and given an opening for a point, it's as difficult finding an opening to the negation of the point as to any other point. However, I do think there's some reduction in security because it's more likely to find a collision with a point and it's negation.

Done.